data_model/src/events/data.rs

```diff
@@ -0,0 +1,498 @@
+                    .as_ref()
+                    .map_or(false, |filter_id| account_id.domain_id == *filter_id),
+                Entity::AssetDefinition(asset_definition_id) => {
+                    opt.as_ref().map_or(false, |filter_id| {
+                        asset_definition_id.domain_id == *filter_id
+                    })
+                }
+                Entity::Asset(asset_id) => opt.as_ref().map_or(false, |filter_id| {
+                    asset_id.account_id.domain_id == *filter_id
+                        || asset_id.definition_id.domain_id == *filter_id
+                }),
+                Entity::Domain(id) => opt.as_ref().map_or(true, |filter_id| id == filter_id),
+                _ => false,
+            },
+            Self::Peer(opt) => match entity {
+                Entity::Peer(peer_id) => {
+                    opt.as_ref().map_or(true, |filter_id| peer_id == filter_id)
+                }
+                _ => false,
+            },
+        }
+    }
+}
+
+impl StatusFilter {
+    fn apply(self, status: Status) -> bool {
+        match self {
+            Self::Created => Status::Created == status,
+            Self::Updated(opt) => match status {
+                Status::Updated(detail) => {
+                    opt.map_or(true, |filter_detail| detail == filter_detail)
+                }
+                _ => false,
+            },
+            Self::Deleted => Status::Deleted == status,
+        }
+    }
+}
+
+impl From<Event> for Vec<Event> {
+    fn from(src: Event) -> Self {
+        vec![src]
+    }
+}
+
+impl From<MetadataUpdated> for Status {
+    fn from(src: MetadataUpdated) -> Self {
+        Self::Updated(src.into())
+    }
+}
+
+impl From<AssetUpdated> for Status {
+    fn from(src: AssetUpdated) -> Self {
+        Self::Updated(src.into())
+    }
+}
+
+mod world {
+    use crate::prelude::*;
+
+    impl From<Register<Peer>> for DataEvent {
+        fn from(src: Register<Peer>) -> Self {
+            Self::new(src.object.id, DataStatus::Created)
+        }
+    }
+
+    impl From<Unregister<Peer>> for DataEvent {
+        fn from(src: Unregister<Peer>) -> Self {
+            Self::new(src.object_id, DataStatus::Deleted)
+        }
+    }
+
+    impl From<Register<Domain>> for DataEvent {
+        fn from(src: Register<Domain>) -> Self {
+            Self::new(src.object.id, DataStatus::Created)
+        }
+    }
+
+    impl From<Unregister<Domain>> for DataEvent {
+        fn from(src: Unregister<Domain>) -> Self {
+            Self::new(src.object_id, DataStatus::Deleted)
+        }
+    }
+
+    #[cfg(feature = "roles")]
+    impl From<Register<Role>> for DataEvent {
+        fn from(src: Register<Role>) -> Self {
+            Self::new(src.object.id, DataStatus::Created)
+        }
+    }
+
+    #[cfg(feature = "roles")]
+    impl From<Unregister<Role>> for DataEvent {
+        fn from(src: Unregister<Role>) -> Self {
+            Self::new(src.object_id, DataStatus::Deleted)
+        }
+    }
```

Comment:
looking at these, maybe some declarative macro might have been written? but it would prob just reduce readability

Reply:
Makes sense, I'll make it macro

Reply:
I'm not sure if that's even possible

Reply:
I decided not to use macros this time
because it seems to require quite a few declarations